apm-agent-plugins/apm-servlet-plugin/src/test/java/co/elastic/apm/agent/servlet/TestRequestBodyCapturing.java

```diff
@@ -0,0 +1,263 @@
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.junit.jupiter.params.provider.ValueSource;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.MediaType;
+import org.springframework.mock.web.MockFilterChain;
+import org.springframework.mock.web.MockHttpServletRequest;
+import org.springframework.mock.web.MockHttpServletResponse;
+
+import javax.annotation.Nonnull;
+import javax.servlet.ServletException;
+import javax.servlet.ServletInputStream;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletRequestWrapper;
+import javax.servlet.http.HttpServletResponse;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.stream.Stream;
+
+import static org.assertj.core.api.Java6Assertions.assertThat;
+import static org.mockito.Mockito.when;
+
+class TestRequestBodyCapturing extends AbstractInstrumentationTest {
+
+    private static final byte[] BUFFER = new byte[1024];
+    private InputStreamConsumer streamConsumer;
+    private InputStreamCloser streamCloser;
+    private WebConfiguration webConfiguration;
+    private MockFilterChain filterChain;
+
+    static Stream<InputStreamConsumer> streamConsumers() {
+        return Stream.of(
+            InputStream::read,
+            is -> is.read(BUFFER),
+            is -> is.read(BUFFER, 0, BUFFER.length),
+            is -> is.readLine(BUFFER, 0, BUFFER.length),
+            is -> {
+                is.readNBytes(BUFFER, 0, BUFFER.length);
+                return -1;
+            },
+            is -> {
+                is.readAllBytes();
+                return -1;
+            }
+        );
+    }
+
+    @BeforeEach
+    void setUp() {
+        webConfiguration = tracer.getConfig(WebConfiguration.class);
+        when(webConfiguration.getCaptureBody()).thenReturn(WebConfiguration.EventType.ALL);
+        filterChain = new MockFilterChain(new HttpServlet() {
+            @Override
+            protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException {
+                final ServletInputStream is = req.getInputStream();
+                int read;
+                do {
+                    read = streamConsumer.read(is);
+                } while (read != -1);
+                streamCloser.close(is);
+            }
+        });
+        streamConsumer = is -> is.readLine(BUFFER, 0, BUFFER.length);
+        streamCloser = InputStream::close;
+
+    }
+
+    @ParameterizedTest
+    @MethodSource("streamConsumers")
+    void testReadTextPlain(InputStreamConsumer consumer) throws Exception {
+        streamConsumer = consumer;
+        executeRequest(filterChain, "foo".getBytes(StandardCharsets.UTF_8), "text/plain");
```

Comment:
Maybe test a two-line body?

Reply:
right, that would yield a better coverage of the `readLine` method